### Reduction operation: the sum of the numbers in the range [0, a)

FUNCIÓN EJEMPLO

In [3]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.018224477767944336 seconds
16.6 ms ± 177 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



3.2 A)

In [2]:
import time

# Crear la lista de números del rango [0, 10**6)
value = 10**6
numeros = list(range(value))

# Método a) Usando un bucle for
initialTime_for = time.time()
suma_for = 0
for numero in numeros:
    suma_for += numero
finalTime_for = time.time()

# Método b) Usando la función sum
initialTime_sum = time.time()
suma_sum = sum(numeros)
finalTime_sum = time.time()

# Mostrar resultados
print("Método a) Usando bucle for:")
print(f"Suma: {suma_for}, Tiempo tomado: {finalTime_for - initialTime_for} segundos\n")

print("Método b) Usando función sum:")
print(f"Suma: {suma_sum}, Tiempo tomado: {finalTime_sum - initialTime_sum} segundos\n")


# Usar %timeit para medir tiempos en IPython (si estás en un entorno compatible)
try:
    get_ipython().run_line_magic('timeit', '-r 2 sum(numeros)')
except NameError:
    print("Las operaciones mágicas de %timeit sólo funcionan en entornos IPython.")

Método a) Usando bucle for:
Suma: 499999500000, Tiempo tomado: 0.4565746784210205 segundos

Método b) Usando función sum:
Suma: 499999500000, Tiempo tomado: 0.02854609489440918 segundos

29.2 ms ± 181 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)


3.2 B)

In [3]:
import time
import numpy as np

# Crear la lista de números
value = 10**6
lista = list(range(value))

# Convertir la lista a un array de NumPy
array = np.array(lista)

# Sumar los elementos usando un bucle for
start_time = time.time()
suma_for = 0
for elemento in array:
    suma_for += elemento
end_time = time.time()
print(f"Tiempo con bucle for (array de NumPy): {end_time - start_time:.6f} segundos")

# Sumar los elementos usando numpy.sum()
start_time = time.time()
suma_numpy = np.sum(array)
end_time = time.time()
print(f"Tiempo con numpy.sum(): {end_time - start_time:.6f} segundos")

# Verificar que ambos resultados son iguales
assert suma_for == suma_numpy, "Los resultados de las sumas no coinciden"
print(f"\nSuma total: {suma_numpy}")

Tiempo con bucle for (array de NumPy): 0.839921 segundos
Tiempo con numpy.sum(): 0.002087 segundos

Suma total: 499999500000


3.2 C)

Los resultados muestran que:

1. El código original: 
es notablemente rápido debido a su implementación optimizada en Python, con tiempos cercanos a 0.018 segundos.

2. El uso de listas:
El bucle for resulta considerablemente más lento (0.457 segundos) debido a la sobrecarga de Python al iterar elemento por elemento.
La función sum es significativamente más eficiente (0.029 segundos), ya que está optimizada para operaciones de reducción en listas.

3. El uso de NumPy:
El bucle for es aún más lento (0.84 segundos) porque la iteración en arrays de NumPy introduce una mayor sobrecarga que en listas nativas.
La función numpy.sum es increíblemente rápida (0.0021 segundos), ya que realiza operaciones directamente en código compilado C, aprovechando instrucciones de bajo nivel y optimizaciones vectoriales.